### PR TITLE
Logic cleanups for `basic_string`

### PR DESCRIPTION
**Pull request by achabense**
This is a reorganized version of #3830.
- Various logic cleanups to make `basic_string` easier to maintain.
- Fixes a bug in `operator=(const&)` which will break invariant.

Thanks to @StephanTLavavej, @fsb4000 and @frederick-vs-ja for review in the original pr!

### CODE REVIEW

File: stl/inc/xstring [hunk 1]
```diff
@@ -3165,21 +3159,24 @@ public:
                 auto&& _Right_alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Right_al);
                 _Container_proxy_ptr<_Alty> _New_proxy(_Right_alproxy, _Leave_proxy_unbound{}); // throws
 
-                if (_Right._Mypair._Myval2._Large_string_engaged()) {
-                    const auto _New_size     = _Right._Mypair._Myval2._Mysize;
-                    const auto _New_capacity = _Calculate_growth(_New_size, 0, _Right.max_size());
```
Comment (achabense) [inline stl/inc/xstring:3170]: This part was broken. It can cause allocation smaller than `_SMALL_STRING_CAPACITY`. (e.g. consider when `large_string_engaged()`, but `size() == 0`)

File: stl/inc/xstring [hunk 2]
```diff
@@ -3165,21 +3159,24 @@ public:
...
                     _Mypair._Myval2._Bx._Ptr = _New_ptr;
-                    _Mypair._Myval2._Mysize  = _New_size;
+                    _Mypair._Myval2._Mysize  = _Right_size;
                     _Mypair._Myval2._Myres   = _New_capacity;
                 } else {
-                    _Copy_assign_val_from_small(_Right);
+                    _Tidy_deallocate();
+                    _Traits::copy(_Mypair._Myval2._Bx._Buf, _Right_ptr, _Right_size + 1);
+                    _Mypair._Myval2._Mysize = _Right_size;
+                    _Mypair._Myval2._Myres  = _SMALL_STRING_CAPACITY;
```
Comment (achabense) [inline stl/inc/xstring:3179]: I think dropping `_Memcpy_val_from` optimization (was in `_Copy_assign_val_from_small`) is ok here, as this branch looks rarely used.
Otherwise we have to deal with three cases:
right is in small mode(res) and in non-constexpr context (and `!defined(_INSERT_STRING_ANNOTATION)`) -> can do `_Memcpy_val_from`
right is small(size) -> copy str without allocation
right is large(size) -> allocate and copy str

File: stl/inc/xstring [hunk 3]
```diff
@@ -3165,21 +3159,24 @@ public:
...
                 _Container_proxy_ptr<_Alty> _New_proxy(_Right_alproxy, _Leave_proxy_unbound{}); // throws
 
-                if (_Right._Mypair._Myval2._Large_string_engaged()) {
-                    const auto _New_size     = _Right._Mypair._Myval2._Mysize;
-                    const auto _New_capacity = _Calculate_growth(_New_size, 0, _Right.max_size());
+                const auto _Right_size        = _Right._Mypair._Myval2._Mysize;
+                const _Elem* const _Right_ptr = _Right._Mypair._Myval2._Myptr();
+                if (_Entails_large_string(_Right_size)) {
+                    const auto _New_capacity =
+                        _Calculate_growth(_Right_size, _SMALL_STRING_CAPACITY, _Right.max_size());
```
Comment (achabense) [inline stl/inc/xstring:3166]: [won't be included in this pr]
If this fix is adopted, the second parameter of `_Calculate_growth(x,y,z)` will be uniformly `_SMALL_STRING_CAPACITY` and further rectification will be possible.
There is also a strength use of `_Calculate_growth(newsize)` in `_Construct_from_iter`. According to impl logic it is performing capacity*=1.5 logic. What about replacing the grow logic here with `_Reallocate_grow_by` or `push_back`?https://github.com/microsoft/STL/blob/40640c6df9754ea031f8df7c720b89714d19df97/stl/inc/xstring#L2712

File: stl/inc/xstring [hunk 4]
```diff
@@ -2290,8 +2294,10 @@ public:
     };
     _Bxty _Bx;
 
-    size_type _Mysize = 0; // current length of string
-    size_type _Myres  = 0; // current storage reserved for string
+    // invariant: _Myres >= _Mysize, and _Myres >= _SMALL_STRING_CAPACITY (after string's construction)
+    // both _Mysize and _Myres doesn't take account of the extra null terminator
+    size_type _Mysize = 0; // current length of string (size)
+    size_type _Myres  = 0; // current storage reserved for string (capacity)
```
Comment (achabense) [inline stl/inc/xstring:2300]: [won't be included in this pr] suggestion: we can make `_Myres` initialized to `_SMALL_STRING_CAPACITY` to make `_Myres >= _SMALL_STRING_CAPACITY` throughout the lifetime of `_String_val`.

File: stl/inc/xstring [hunk 5]
```diff
@@ -2625,9 +2650,9 @@ private:
         auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Al);
         _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _My_data);
 
-        if (_Count < _BUF_SIZE) {
+        if (!_Entails_large_string(_Count)) {
             _My_data._Mysize = _Count;
-            _My_data._Myres  = _BUF_SIZE - 1;
+            _STL_INTERNAL_CHECK(_My_data._Myres == _SMALL_STRING_CAPACITY);
```
Comment (achabense) [inline stl/inc/xstring:2655]: Hmm forgot to revert this change...

File: stl/inc/xstring [hunk 6]
```diff
@@ -2369,6 +2369,13 @@ private:
     static constexpr size_type _ALLOC_MASK            = _Scary_val::_ALLOC_MASK;
     static constexpr size_type _SMALL_STRING_CAPACITY = _Scary_val::_SMALL_STRING_CAPACITY;
 
+    _NODISCARD static _CONSTEXPR20 bool _Entails_large_string(size_type _Size_or_capacity) noexcept {
+        // Use cases:
+        // 1. determine whether large mode is used for a given capacity.
```
Comment (achabense) [inline stl/inc/xstring:2376]: As case 1 is not used often and shares exactly the same intention as `_Large_string_engaged`(all replaced in [this commit](https://github.com/microsoft/STL/pull/3862/commits/8ca3f1c499a9725336ebc039a7a478556cb63e7d)), I think we'd better add comments like `// large string engaged` for them.

File: stl/inc/xstring [hunk 7]
```diff
@@ -2589,19 +2607,26 @@ public:
...
+            for (size_type _Idx = 0; _Idx < _Capacity + 1; ++_Idx) {
                 _STD construct_at(_Ptr + _Idx);
             }
         }
-#else // ^^^ C++20-or-later / pre-C++20 vvv
-        (void) _Ptr;
-        (void) _Size;
 #endif // _HAS_CXX20
+        return _Fancy_ptr;
+    }
```
Comment (achabense) [inline stl/inc/xstring:2624]: As to `_Allocate_at_least_helper` introduced in #3864, I think we can make `_Allocate_for_capacity` apply `_Allocate_at_least_helper` logic, then most of the rest part in this pr will work without change. The only place that needs extra work will be an allocation in the interaction part with `basic_stringbuf`.

Side note:
In the past we have to "explicitly" take care of the extra null terminator for both size(when copy) and capacity(when allocate). If `_(De)allocate_for_capacity` is applied, only size will need manual `+1` logic.

File: stl/inc/xstring [hunk 8]
```diff
@@ -2932,7 +2950,7 @@ public:
     struct _Released_buffer {
         pointer _Ptr;
         size_type _Size;
-        size_type _Res;
+        size_type _Res; // real allocation size
```
Comment (achabense) [inline stl/inc/xstring:2953]: I think `_Res` is a terrible name as it is conflicting with that of `_String_val::_Myres`. In what condition can we change the name of a data member?

**Comment by StephanTLavavej** [comment]
There are non-trivial merge conflicts with #3864, the most significant being between `_Allocate_for_capacity` and `_Allocate_at_least_helper`. Do you have time to rebase and reorganize this one more time?

**Comment by achabense** [comment]
I could try to solve the conflict. After the change, `_Allocate_for_capacity` will include `_Start_element_lifetimes` and `_Allocate_at_least_helper` logic.

However, before that, I need to confirm that introduction of `_Allocate_for_capacity` is a good decision.  
1. The terminating null still need to be cared about explicitly in during copy. 
2. `_Allocate_at_least_helper` still need to be used explicitly in the intedraction part with `basic_stringbuf`.

**Comment by fsb4000** [comment]
I think this is a compiler bug and your PR is fine.
From CI:
```
Build step failed unexpectedly.
Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.37.32820\bin\HostX64\ARM\cl.EXE" "D:\a\_work\1\s\tests\std\tests\Dev09_181509_tr1_inf_loop_uniform_int_ull\test.cpp" "-ID:\build\out\inc" "-ID:\a\_work\1\s\llvm-project\libcxx\test\support" "-ID:\a\_work\1\s\tests\std\include" "/nologo" "/Od" "/W4" "/w14061" "/w14242" "/w14265" "/w14582" "/w14583" "/w14587" "/w14588" "/w14749" "/w14841" "/w14842" "/w15038" "/w15214" "/w15215" "/w15216" "/w15217" "/w15262" "/sdl" "/WX" "/D_ENABLE_STL_INTERNAL_CHECK" "/bigobj" "/FIforce_include.hpp" "/w14365" "/w15267" "/D_ENFORCE_FACET_SPECIALIZATIONS=1" "/D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER" "/w14640" "/Zc:threadSafeInit-" "/EHsc" "/MDd" "/D_ITERATOR_DEBUG_LEVEL=2" "/std:c++20" "/permissive-" "-FeD:\build\tests\std\tests\Dev09_181509_tr1_inf_loop_uniform_int_ull\Output\09\Dev09_181509_tr1_inf_loop_uniform_int_ull.exe" "-link" "-LIBPATH:D:\build\out\lib\arm" "-LIBPATH:C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.37.32820\lib\arm" "/MANIFEST:EMBED"
Exit Code: 57005
Standard Output:
--
test.cpp
D:\a\_work\1\s\tests\std\tests\Dev09_181509_tr1_inf_loop_uniform_int_ull\test.cpp : fatal error C1001: Internal compiler error.
(compiler file 'D:\a\_work\1\s\src\vctools\Compiler\Utc\src\p2\main.c', line 234)
 To work around this problem, try simplifying or changing the program near the locations listed above.
If possible please provide a repro here: https://developercommunity.visualstudio.com 
Please choose the Technical Support command on the Visual C++ 
 Help menu, or open the Technical Support help file for more information
```
I don't have an ARM laptop to run the test on and try to reduce.
@StephanTLavavej Should I create a Devcom ticket as is, or would you be able to help the FE devs by reducing the test?

**Comment by StephanTLavavej** [comment]
@fsb4000

> ```
> Command: "C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.37.32820\bin\HostX64\ARM\cl.EXE"
> D:\a\_work\1\s\tests\std\tests\Dev09_181509_tr1_inf_loop_uniform_int_ull\test.cpp : fatal error C1001: Internal compiler error.
> (compiler file 'D:\a\_work\1\s\src\vctools\Compiler\Utc\src\p2\main.c', line 234)
> ```
> I don't have an ARM laptop to run the test on and try to reduce.

You don't need one - the CI is cross-compiling from x64 to ARM (look at the cl.exe path saying `HostX64\ARM`), and it's ICEing during the compilation. In fact, we don't have ARM[64]-based CI VMs yet, which is why the tests are only built and not run.

It helps to know the cross-compiler terminology:

* The "build" machine is where the compiler is built. For MSVC, this happens in the MSVC build labs (which are x64), so it usually isn't mentioned.
* The "host" machine is where the compiler binary runs on the user-programmer's machine.
* The "target" machine is what the compiler is generating binaries for, the end-user's machine.

> @StephanTLavavej Should I create a Devcom ticket as is, or would you be able to help the FE devs by reducing the test?

This is actually a BE bug (indicated by `Utc\src\p2` in the ICE - UTC is the internal name for the backend, the Unified Tuple Compiler), indicating that it's related to the optimizations used by `Dev09_181509_tr1_inf_loop_uniform_int_ull`'s `fast_matrix`.

It should be fairly straightforward to extract a preprocessed repro locally. You'll need the ARM tools installed, and then the incantations (for 32-bit ARM) are (modify paths for your machine):

```
"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Auxiliary\Build\vcvarsall.bat" x64_arm
set STL=D:\GitHub\STL
pushd %STL%
cmake -G Ninja -S . -B out\arm -DTESTS_BUILD_ONLY=ON
ninja -C out\arm
pushd out\arm
python tests\utils\stl-lit\stl-lit.py %STL%\tests\std\tests\Dev09_181509_tr1_inf_loop_uniform_int_ull -o testing_arm.log
python %STL%\tools\scripts\print_failures.py testing_arm.log
```

**Comment by fsb4000** [comment]
Ok, I will try to reduce.

File: stl/inc/xstring [hunk 9]
```diff
@@ -3992,7 +3992,7 @@ public:
             return;
         }
 
-        const size_type _Target_capacity = (_STD min)(_My_data._Mysize | _ALLOC_MASK, max_size());
+        size_type _Target_capacity = (_STD min)(_My_data._Mysize | _ALLOC_MASK, max_size());
```
Comment (achabense) [inline stl/inc/xstring:3995]: Hmm, perhaps irrelevant to the test_failure, but I realize that after the change `_Allocate_for_capacity` should not apply to `shrink_to_fit` here.

Comment (achabense) [reply stl/inc/xstring:3995]: On second thought, I think this change is not as harmful as it appears, as `shrink_to_fit` `"is a non-binding request to reduce capacity() to size()"`, and in my understanding a replacement of `allocate` to `allocate_at_least` will not result in extra allocation(maybe I'm wrong here) - it just reflect a situation that the string can end up `taking up` more space(whether usable by the string) after "shrinking"

**Comment by fsb4000** [comment]
Almost every change fixes the test, this complicates the process of reducing the test.

Current results:

изображение


Build command reduced to
```
"C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.37.32820\bin\HostX64\ARM\cl.EXE" /c main.cpp "-IC:\Dev\STL\out\arm\out\inc" "/nologo" "/Od" "/sdl" "/D_ENABLE_STL_INTERNAL_CHECK" "/D_ENFORCE_FACET_SPECIALIZATIONS=1"  "/EHsc" "/MDd" "/std:c++20"
```

The test reduced to
```c++
struct TestEnvironmentPreparer {
    TestEnvironmentPreparer() {
    }

    bool succeeded() const {
        return true;
    }
};

const TestEnvironmentPreparer test_environment_preparer{};

#include <algorithm>
#include <execution>
#include <random>
#include <vector>

using namespace std;
using namespace std::execution;

constexpr auto int16_min  = -32768;
constexpr auto int16_max  = 32767;
constexpr auto int32_min  = -2147483647-1;
constexpr auto int32_max  = 2147483647;
constexpr auto int64_min  = -9223372036854775807-1;
constexpr auto int64_max  = 9223372036854775807;
constexpr auto uint16_max = 65535;
constexpr auto uint32_max = 4294967295;
constexpr auto uint64_max = 18446744073709551615;

constexpr int N = 1;

template <typename URNG, typename T, T A, T B>
void microtest() {
    URNG engine;
    uniform_int_distribution<T> dist(777, 777);

    const T val = dist(engine);
    for (int i = 0; i < N; ++i) {

    }

}

using fp_t = void (*)();
template <typename URNG>
void add_tests(vector<fp_t>& tests) {
    // Test DevDiv-83370 "uniform_int_distribution isn't uniform".
    tests.insert(
        tests.end(), {
                         microtest<URNG, long long, int16_min, int16_max>,
                         microtest<URNG, long long, int16_min + 1, int16_max>,
                         microtest<URNG, long long, int16_min, int16_max - 1>,
                         microtest<URNG, long long, int16_min + 2, int16_max - 3>,
                         microtest<URNG, short, -4, 4>,
                         microtest<URNG, short, 11, 22>,
                         microtest<URNG, short, -44, -33>,
                         microtest<URNG, short, 5, 8>,
                         microtest<URNG, short, -7, -7>,

                         microtest<URNG, unsigned short, 0, uint16_max>,
                         microtest<URNG, unsigned short, 1, uint16_max>,
                         microtest<URNG, unsigned short, 0, uint16_max - 1>,
                         microtest<URNG, unsigned short, 2, uint16_max - 3>,
                         microtest<URNG, unsigned short, 123, 456>,
                         microtest<URNG, unsigned short, 1000, 1001>,
                         microtest<URNG, unsigned short, 777, 777>,

                         microtest<URNG, int, int32_min, int32_max>,
                         microtest<URNG, int, int32_min + 2, int32_max - 3>,
                         microtest<URNG, unsigned, 0, uint32_max>,
                         microtest<URNG, unsigned, 2, uint32_max - 3>,
                         microtest<URNG, long long, int64_min, int64_max>,
                         microtest<URNG, long long, int64_min + 2, int64_max - 3>,
                         microtest<URNG, unsigned long long, 0, uint64_max>, // Test DDB-181509 "TR1 VC9 SP1: Infinite loop in
                                                                   // uniform_int<unsigned long long>::_Eval()".
                         microtest<URNG, unsigned long long, 2, uint64_max - 3>,

                         microtest<URNG, int, -4, 4>,
                         microtest<URNG, int, 11, 22>,
                         microtest<URNG, int, -44, -33>,
                         microtest<URNG, int, 5, 6>,
                         microtest<URNG, int, -7, -7>,

                         microtest<URNG, unsigned, 123, 456>,
                         microtest<URNG, unsigned, 1000, 1001>,
                         microtest<URNG, unsigned, 777, 777>,

                         microtest<URNG, long long, -4, 4>,
                         microtest<URNG, long long, 11, 22>,
                         microtest<URNG, long long, -44, -33>,
                         microtest<URNG, long long, 5, 6>,
                         microtest<URNG, long long, -7, -7>,

                         microtest<URNG, unsigned long long, 123, 456>,
                         microtest<URNG, unsigned long long, 1000, 1001>,
                         microtest<URNG, unsigned long long, 777, 777>,
                     });
}

class Wacky {
public:
    Wacky() : mt(0) {}

    typedef unsigned long result_type;

    static result_type min() {
        return 0;
    }

    static result_type max() {
        return 90000;
    }

    result_type operator()() {
        for (;;) {
            const result_type ret = mt() & 0x1FFFFUL;

            if (ret >= min() && ret <= max()) {
                return ret;
            }
        }
    }

private:
    mt19937 mt;
};

int main() {
    vector<fp_t> tests;

    add_tests<Wacky>(tests);
    add_tests<mt19937>(tests);
    add_tests<mt19937_64>(tests);


    for_each(par, tests.begin(), tests.end(), [](fp_t fn) { fn(); });
}
```

For example:
1) if we move
```
struct TestEnvironmentPreparer {
    TestEnvironmentPreparer() {
    }

    bool succeeded() const {
        return true;
    }
};

const TestEnvironmentPreparer test_environment_preparer{};
```
after
```
using namespace std::execution;
```
the test works.

2) Or if we change
```
TestEnvironmentPreparer() {
}
```
to
```
TestEnvironmentPreparer() = default;
```
the tests works.

3) Or if we change
```
const T val = dist(engine);
```
to
```
T val = dist(engine);
```
the tests works.

4) Or if we remove the for
```
for (int i = 0; i < N; ++i) {
}
```
the test works.

5) Or if we remove the
```
microtest<URNG, unsigned long long, 777, 777>,
```
from the vector
the test works

6) Or if we add additional function to the vector.
change
```
microtest<URNG, unsigned long long, 777, 777>,
```
to
```
microtest<URNG, unsigned long long, 777, 777>,
microtest<URNG, unsigned long long, 777, 778>,
```
the test works

7) If we change
```
for_each(par, tests.begin(), tests.end(), [](fp_t fn) { fn(); });
```
to normal range-based for then the test works.

8) If we change
```
result_type operator()() {
        for (;;) {
            const result_type ret = mt() & 0x1FFFFUL;

            if (ret >= min() && ret <= max()) {
                return ret;
            }
        }
    }
```
to
```
result_type operator()() {
    return mt() & 0x1FFFFUL;
}
```
the test works.

9) Any two combination  of `add_tests` work. ICE only when we have 3 `add_tests`.
```
    add_tests<Wacky>(tests);
    add_tests<mt19937>(tests);
    add_tests<mt19937_64>(tests);
```

**Comment by StephanTLavavej** [comment]
@fsb4000 That looks sufficiently reduced to be able to file a DevCom bug. The BE devs just need a self-contained repro - they are used to working with enormous link repros, so a source file that includes only the STL will be delightful to them.

**Comment by fsb4000** [comment]
I've created Devcom-10419500

**Comment by StephanTLavavej** [comment]
I am going to force-push this branch to clean up the history, and also resolve the test failure. Please hold off on pushing any changes until I am done here, thanks!

File: stl/inc/xstring (hunk 1 above)
Comment (StephanTLavavej) [reply stl/inc/xstring]: **GREAT CATCH!!!** :heart_eyes_cat: I verified that this caused observably damaged behavior.

File: stl/inc/xstring (hunk 2 above)
Comment (StephanTLavavej) [reply stl/inc/xstring:3190]: Yeah, this is in the highly obscure POCCA codepath for non-equal allocators. Correctness is fine, we don't need to micro-optimize this case.

File: stl/inc/xstring (hunk 8 above)
Comment (StephanTLavavej) [reply stl/inc/xstring:2965]: Messing with *types* poses grave ABI concerns. For ABI-locked code (i.e. older than `/std:c++latest`), we can't ever do this for data members of types that the user can directly create or derive from.

We can mess with types when they are purely internal - e.g. a struct that is used only as a local variable within a function. In such cases, we should rename the type to avoid ODR issues.

In certain cases we can also rename internal types when their manipulation is purely through virtuals; this happened with `shared_ptr`'s control blocks, for example.

Messing with names will break visualizers (if present) and we shouldn't do it without a very good reason.

With `_Released_buffer::_Res`, maybe we can rename it.

File: stl/inc/xstring (hunk 9 above)
Comment (StephanTLavavej) [reply stl/inc/xstring:4005]: I don't think we should mess with `shrink_to_fit`'s behavior, and have pushed a commit to imitate `vector`'s exact policy.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
As a special exception to our usual way of working, I've force-pushed the branch to rebase it on `main` and linearize the history. I verified (with `git diff`) that this results in the exact same state. I fused commits (e.g. `_NODISCARD`) that were logically grouped together, ironed out commits that were purely addressing earlier mistakes, and decomposed a couple of commits into finer-grained ones (e.g. `auto`, `_CONSTEXPR20`).

**THANK YOU** for the fine-grained commits, they were an enormous help in reviewing. After linearization, I was able to see the correctness of each transformation and I have high confidence in the resulting state. Overall I think that this is a great cleanup and makes the complicated `basic_string` code more robust for future maintenance - particularly avoiding off-by-one errors around the SSO, and extracting the `allocate_at_least` and start-element-lifetimes complexity that has recently accreted.

Finally, I added commits to address the minor issues I found while reviewing:

* Serially run the tests in `Dev09_181509_tr1_inf_loop_uniform_int_ull`.
  + This should avoid the mysterious compiler back-end ICE when cross-compiling from x64 to ARM. This test originally (6+ years ago) tested so many cases, it was reasonable to run them in parallel. Those cases got chopped away over the years (last major change was in 2017), and the resulting test runs in ~4 seconds serially. Because we test so many configurations in parallel, there is no advantage to making this test run parallel inside itself, so I've converted it to the C++14 matrix, dropped the use of parallel `for_each()`, and removed the `/clr` workaround (for crashes with parallel algorithms).
* Fix comment grammar.
* In `_Deallocate_for_capacity`, pass `_Old_ptr` by value.
  + We (and the Standard) conventionally pass potentially-fancy pointers by value; they are expected to be cheap to copy.
* Drop redundant `public:` after the removal of `_Copy_assign_val_from_small`.
* Add `test_string_copy_assign_pocca_sso()`.
  + I verified that this fails before your fix, and passes afterwards. The existing coverage didn't exercise the SSO transition enough. It was simplest to just add a new case rather than trying to alter the existing coverage, and this test conveniently had the POCCA logic ready to go.
* Restore `shrink_to_fit` to allocate exactly.
  + We should behave like `vector` and avoid the `allocate_at_least` optimization here. Let's avoid surprising users; they already worry about `shrink_to_fit`'s non-binding nature.

File: stl/inc/xstring [hunk 10]
```diff
@@ -2588,19 +2606,39 @@ public:
     }
 
 private:
-    static constexpr void _Start_element_lifetimes(_Elem* const _Ptr, const size_type _Size) {
+    enum class _Reallocation_policy { _At_least, _Exactly };
```
Comment (achabense) [inline stl/inc/xstring:2609]: Thanks for your help! As to the new changes, I think `_Allocation_policy` will be a better name here, as `_Allocate_for_capacity` only involves allocation logic. (vector uses `_Reallocation_policy` for `_Reallocate`)

**Review by StephanTLavavej — APPROVED** [review]

File: stl/inc/xstring (hunk 10 above)
Comment (StephanTLavavej) [reply stl/inc/xstring]: You're welcome! And great catch, I wasn't thinking enough as I was copy-pasting that code. I've pushed a rename.

File: stl/inc/xstring [hunk 11]
```diff
@@ -4001,22 +3997,21 @@ public:
...
         }
 
-        if (_My_data._Mysize < _BUF_SIZE) {
+        if (!_Entails_large_string(_My_data._Mysize)) {
             _Become_small();
             return;
         }
 
-        const size_type _Target_capacity = (_STD min)(_My_data._Mysize | _ALLOC_MASK, max_size());
+        size_type _Target_capacity = (_STD min)(_My_data._Mysize | _ALLOC_MASK, max_size());
```
Comment (achabense) [inline stl/inc/xstring:4005]: (no changes requested) As `_Exactly` mode doesn't really require input capacity to be a referrence, I think we can restore constness for `_Target_capacity` in the future. Also here in `<vector>`:
https://github.com/microsoft/STL/blob/0fe653a9d6efd2ac0ad07825804436672dbffe3b/stl/inc/vector#L1705-L1706

**Review by CaseyCarter — APPROVED** [review]
I'll verify and push a change to address my comments, which are mostly naming nits.

File: stl/inc/xstring (hunk 8 above)
Comment (CaseyCarter) [reply stl/inc/xstring]: Renamed to `_Actual_allocation_size`.

File: stl/inc/xstring (hunk 6 above)
Comment (StephanTLavavej) [reply stl/inc/xstring]: This is moot after Casey expanded this function at each callsite.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks again for this maintainability overhaul and runtime correctness fix! :heart_eyes_cat: :magic_wand: :rocket:

File: stl/inc/xstring (hunk 8 above)
Comment (achabense) [reply stl/inc/xstring]: I've checked the changes to `mutex` in pr #3824. It looks I can move `monostate` back to `<variant>`  by dropping reliance from `basic_format_arg`.
I think defining `monostate` in `<xutility>` is undesirable, as the standard require it to be defined in `<variant>`, and all supporting utilities (compare/hash) and use cases remain in `<variant>`.
This is the change: https://github.com/microsoft/STL/compare/main...achabense:STL:_Cleanups, could you help me check whether it is a safe approach? (There is another place I'm not confident about in `<exception>` in the branch)

Comment (StephanTLavavej) [reply stl/inc/xstring]: That seems reasonable, although I would simply direct-init `_Monostate _No_state{};`.

Changing `_With_nested`'s template parameters should affect mangling, making it a safe change (internal class that users can't directly store). I think it would be better to rename it with `_v2` though, just to be extra safe.
